### ISSUE TITLE
How to use detect.py and val.py to evaluate my model?

### ISSUE THREAD

**Issue by Zhiying-Li-dot** [ultralytics/yolov5#5512]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I change the yolov5 network by adding some tricks.I can use the train.py to train my network, but how to use val.py and detect.py to evaluate my model.Could you please tell me how to do?Thanks a lot.

**Comment by Zengyf-CVer** [comment]
@lizhiying-2019 
In fact, this is very simple, as long as you set the relevant parameters in val.py and detect.py.
### val.py
https://github.com/ultralytics/yolov5/blob/0155548384e90597f140560004364633430070cd/val.py#L302-L307
### detect.py
https://github.com/ultralytics/yolov5/blob/0155548384e90597f140560004364633430070cd/detect.py#L276-L280

**Comment by glenn-jocher** [comment]
@lizhiying-2019 also see val.py and detect.py files directly for usage examples:
https://github.com/ultralytics/yolov5/blob/17b5f5b974784ab6a4c477a3964b896123fd8808/detect.py#L3-L6

https://github.com/ultralytics/yolov5/blob/17b5f5b974784ab6a4c477a3964b896123fd8808/val.py#L3-L6

**Comment by Zhiying-Li-dot** [comment]
@Zengyf-CVer  @glenn-jocher Yeah,I have noticed the relevant parameters.But there are weights and data, no cfg to load my model. I use the best.pt obtained from training to set the parameters. But when running, the number of network layers does not match my model. So what should I do?Just set the weights parameter with the best.pt obtained from training my model?

**Comment by glenn-jocher** [comment]
@lizhiying-2019 its very simple. You just follow the Usage examples given above. I can't design YOLOv5 any simpler than this.
```bash
python path/to/val.py --data path/to/data.yaml --weights path/to/model.pt
```

**Comment by Zhiying-Li-dot** [comment]
Like this?
 `python val.py --data dataSet/clothes.yaml --weights mymodel/mymodel/weights/best.pt`

**Comment by Zhiying-Li-dot** [comment]
@glenn-jocher Do I just need to do this?But when running, the number of network layers does not match my model.

**Comment by Zhiying-Li-dot** [comment]
![微信截图_20211106102608](https://user-images.githubusercontent.com/75170833/140594770-8f8ad276-7455-4f29-b97b-561fc95d109e.png)
Like this, when I run val.py with my model.pt ，the model layers is 476. But the model layers is 607 when running yolo.py.

**Comment by Zhiying-Li-dot** [comment]
@glenn-jocher How to do?I hava no idea.Could you please tell me?

**Comment by Zengyf-CVer** [comment]
@lizhiying-2019 
This is normal, mainly for the custom model layer, 476.

**Comment by Zengyf-CVer** [comment]
@lizhiying-2019 
You are retraining for a round. Observe whether the number of training layers is 476. If it is, it proves that there is no problem.

**Comment by Zhiying-Li-dot** [comment]
@Zengyf-CVer So just set the weights parameter without model.yaml. Is that right？

**Comment by Zhiying-Li-dot** [comment]
@Zengyf-CVer I am still very confused, why val.py has no cfg to load my modified model.The train.py has cfg to load model, I can know that which model I am training now.But val.py only has weights to load pt file,not has cfg to load my modified model, I don't know which model I am evaluating.That's my real issue.

**Comment by Zengyf-CVer** [comment]
@lizhiying-2019 
The so-called cfg file represents the structure of the training model. Because the model has been trained and stored in pt form, it is enough to use the pt model in val.py, and there is no need to quote the cfg file.

**Comment by Zhiying-Li-dot** [comment]
@Zengyf-CVer Thank you for your answer,it helped me a lot.I got it.

**Comment by MLDavies** [comment]
val.py was working recently, then I broke something. Can anyone translate the error message and point me to what I've done wrong? I'm working on google colab. I haven't altered the val.py file. (clearly I'm playing with a few different things as an exercise)

Thanks in advance for any hints/help. 
![image](https://user-images.githubusercontent.com/34901329/177662093-17324803-419e-42fb-8f5d-59d3c336af27.png)

**Comment by glenn-jocher** [comment]
@MLDavies you have no `train:` field in your dataset yaml.

To train correctly your data must be in YOLOv5 format. Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths), 2) the number of classes `nc` and 3) a list of class `names`:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes
nc: 80  # number of classes
names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
         'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
         'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
         'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
         'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
         'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
         'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
         'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
         'hair drier', 'toothbrush' ]  # class names
```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<img width="800" alt="Image Labels" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>


Good luck 🍀 and let us know if you have any other questions!

**Comment by MLDavies** [comment]
To be clear, train: and val: are pointing to the same images?

**Comment by glenn-jocher** [comment]
@MLDavies in our toy example above yes, we want to verify that the model is capable of overfitting the train set. In a real world scenario you use different val set.
